src/target/riscv/batch.c

```diff
@@ -249,3 +295,37 @@ bool riscv_batch_dmi_busy_encountered(const struct riscv_batch *batch)
 	const uint64_t in = buf_get_u64(field->in_value, 0, field->num_bits);
 	return get_field(in, DTM_DMI_OP) == DTM_DMI_OP_BUSY;
 }
+
+/* Returns the entry in `batch->idle_counts` that defines idle_count for the scan. */
+static idle_count_info_t *find_idle_counts_entry(struct riscv_batch *batch, size_t scan_idx)
+{
+	assert(!list_empty(&batch->idle_counts));
+
+	idle_count_info_t *entry;
+	list_for_each_entry(entry, &batch->idle_counts, list)
+		if (entry->until_scan > scan_idx)
+			break;
+	assert(!list_entry_is_head(entry, &batch->idle_counts, list));
+	return entry;
+}
+
+int riscv_batch_change_idle_used_from_scan(struct riscv_batch *batch, size_t new_idle, size_t scan_idx)
+{
+	idle_count_info_t * const new_entry = malloc(sizeof(*new_entry));
+	if (!new_entry) {
+		LOG_ERROR("Out of memory!");
+		return ERROR_FAIL;
+	}
+	new_entry->until_scan = scan_idx;
+	idle_count_info_t *old_entry = find_idle_counts_entry(batch, scan_idx);
+	/* Add new entry before the old one. */
+	list_add_tail(&new_entry->list, &old_entry->list);
+	assert(new_entry->until_scan < old_entry->until_scan);
+	/* new entry now defines the range until the scan (non-inclusive) */
+	new_entry->idle_count = old_entry->idle_count;
+	/* old entry now defines the range from the scan (inclusive) */
+	old_entry->idle_count = new_idle;
+	LOG_DEBUG("Will use idle == %zu from scan %zu until scan %zu.", old_entry->idle_count,
+			new_entry->until_scan, old_entry->until_scan);
+	return ERROR_OK;
+}
```

Comment:
As also noted above, the function jtag_fill_queue can look directly at the value `r->reset_delays_wait` (and decrement it with each added scan), therefore this function would not be needed.